src/rhash/hash.c

```diff
@@ -1087,6 +1098,397 @@ static int rc_hash_n64(char hash[33], const char* path)
   return rc_hash_finalize(&md5, hash);
 }
 
+static int rc_hash_nintendo_3ds_ncch(md5_state_t* md5, void* file_handle, uint8_t header[0x200], struct AES_ctx* aes)
+{
+  const uint32_t MAX_BUFFER_SIZE_IN_MEDIA_UNITS = MAX_BUFFER_SIZE / 0x200;
+  uint8_t* hash_buffer;
+  int64_t exefs_offset;
+  uint32_t exefs_size;
+
+  exefs_offset = ((uint32_t)header[0x1A3] << 24) | (header[0x1A2] << 16) | (header[0x1A1] << 8) | header[0x1A0];
+  exefs_size = ((uint32_t)header[0x1A7] << 24) | (header[0x1A6] << 16) | (header[0x1A5] << 8) | header[0x1A4];
+
+  if (exefs_size > MAX_BUFFER_SIZE_IN_MEDIA_UNITS)
+    exefs_size = MAX_BUFFER_SIZE_IN_MEDIA_UNITS;
+
+  /* Offset and size are in "media units" (1 media unit = 0x200 bytes) */
+  exefs_offset *= 0x200;
+  exefs_size *= 0x200;
+
+  /* This region is technically optional, but it should always be present for executable content (i.e. games) */
+  if (exefs_offset == 0 || exefs_size == 0)
+    return rc_hash_error("ExeFS was not available");
+
+  /* ASSERT: file position must be +0x200 from start of NCCH (i.e. end of header) */
+  exefs_offset -= 0x200;
+
+  if (aes)
+  {
+    /* This is annoying, we might have to "decrypt" the data in-between the header and ExeFS */
+    /* Luckily this shouldn't be a lot of data */
```

Comment:
The nature of the encryption used here doesn't allow for random access decryption. Each block must be decrypted sequentially as it is laid out in the file, as each block affects the AES state.

Technically I could just hash the encrypted contents of ExeFS (although at the very minimum the NCCH header might need a decryption round, so the bulk of the decryption code is going to still be present regardless) However, in this case it will result in a mismatch compared to the .app file which is actually installed. For upstream Citra (and presumably some future RACitra fork), CIAs are installed, and there user simply selects the installed game in the game list, which internally just launches the .app file installed. This would contrast with BizHawk (and presumably some libretro frontend? I don't know how it handles this) which if handed a CIA file will install it if it isn't installed then internally redirect to the installed .app file, although for hashing purposes all that is seen in the CIA file rather than the .app file (and of course the .app file might be the one hashed anyways if the user decides to just manually launch the .app file, which would work completely fine anyways since that's what the core does for an already installed CIA anyways).